core/src/main/java/org/apache/brooklyn/core/mgmt/classloading/ClassLoaderFromStackOfBrooklynClassLoadingContext.java

```diff
@@ -0,0 +1,137 @@
+ * and the worker wants to have a simple view of the loader to use at any point in time.
+ * For example XStream keeps a class-loader in effect but when deserializing things 
+ * some of those things may define bundles to use. */
+public class ClassLoaderFromStackOfBrooklynClassLoadingContext extends ClassLoader {
+    
+    private final Stack<BrooklynClassLoadingContext> contexts = new Stack<BrooklynClassLoadingContext>();
+    private final Stack<ClassLoader> cls = new Stack<ClassLoader>();
+    private final AtomicReference<Thread> lockOwner = new AtomicReference<Thread>();
+    private ManagementContext mgmt;
+    private ClassLoader currentClassLoader;
+    private AtomicReference<ClassLoaderUtils> currentLoader = new AtomicReference<>();
+    private int lockCount;
+    
+    public ClassLoaderFromStackOfBrooklynClassLoadingContext(ClassLoader classLoader) {
+        setCurrentClassLoader(classLoader);
+    }
+    
+    public void setManagementContext(ManagementContext mgmt) {
+        this.mgmt = checkNotNull(mgmt, "mgmt");
+        currentLoader.set(new ClassLoaderUtils(currentClassLoader, mgmt));
+    }
+
+    @Override
+    protected Class<?> findClass(String name) throws ClassNotFoundException {
+        return currentLoader.get().loadClass(name);
+    }
+    
+    @Override
+    protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
+        return findClass(name);
+    }
+
+    /** Must be accompanied by a corresponding {@link #popClassLoadingContext()} when finished. */
+    @SuppressWarnings("deprecation")
+    public void pushClassLoadingContext(BrooklynClassLoadingContext clcNew) {
+        acquireLock();
+        BrooklynClassLoadingContext oldClc;
+        if (!contexts.isEmpty()) {
+            oldClc = contexts.peek();
+        } else {
+            // TODO xml serializers using this should take a BCLC instead of a CL
+            oldClc = JavaBrooklynClassLoadingContext.create(mgmt, getCurrentClassLoader());
+        }
+        BrooklynClassLoadingContextSequential clcMerged = new BrooklynClassLoadingContextSequential(mgmt, oldClc, clcNew);
+        ClassLoader newCL = ClassLoaderFromBrooklynClassLoadingContext.of(clcMerged);
+        contexts.push(clcMerged);
+        cls.push(getCurrentClassLoader());
+        setCurrentClassLoader(newCL);
+    }
+
+    public void popClassLoadingContext() {
+        synchronized (lockOwner) {
+            releaseXstreamLock();
+            setCurrentClassLoader(cls.pop());
+            contexts.pop();
+        }
+    }
+    
+    private ClassLoader getCurrentClassLoader() {
+        return currentClassLoader;
+    }
+    
+    private void setCurrentClassLoader(ClassLoader classLoader) {
+        currentClassLoader = checkNotNull(classLoader);
+        currentLoader.set(new ClassLoaderUtils(currentClassLoader, mgmt));
+    }
+    
+    protected void acquireLock() {
+        synchronized (lockOwner) {
```

Comment:
I thought I took `xstream` out of the names but I missed at least one!

I'm really not sure what the threading guarantees/requirements are.  It was made just for xstream to support osgi and follows whatever was needed there.  i was amused to discover despite its name it had nothing to do with osgi or xstream!  i thought i needed to refactor it for tests though i ended up doing the tests differents; still the rename felt worth keeping.

my 2p it's good to be reminded of this, there might well be other places such a thing might be useful, and if there is then we'll better think through threading and locking.